src/index.js

```diff
@@ -0,0 +1,250 @@
+import { parseBearer } from './helpers/tokenFormatter';
+import {
+  ENVIRONMENT_IS_REACT_NATIVE,
+  ENVIRONMENT_IS_NODE,
+  ENVIRONMENT_IS_WEB,
+  ENVIRONMENT_IS_WORKER,
+  ERROR_REFRESH_TOKEN_EXPIRED,
+  ERROR_INVALID_CONFIG,
+  STATUS_UNAUTHORIZED,
+  STATUS_OK,
+} from './const';
+
+let config = {
+  prepareRefreshTokenRequest: null,
+  shouldIntercept: null,
+  shouldInvalidateAccessToken: null,
+  getAccessTokenFromResponse: null,
+  setRequestAuthorization: null,
+  onAccessTokenAcquired: null,
+};
+
+let tokens = {
+  accessToken: null,
+  refreshToken: null,
+};
+
+let refreshAccessTokenPromise = null;
+
+if (ENVIRONMENT_IS_REACT_NATIVE) {
+  attach(global);
+} else if (ENVIRONMENT_IS_WORKER) {
+  attach(self);
+} else if (ENVIRONMENT_IS_WEB) {
+  attach(window);
+} else if (ENVIRONMENT_IS_NODE) {
+  attach(global);
+} else {
+  throw new Error('Unsupported environment for fetch-token-intercept');
+}
+
+function attach(env) {
+  if (!env.fetch) {
+    throw Error('No fetch available. Unable to register fetch-token-intercept');
+  }
+
+  // monkey patch fetch
+  env.fetch = (function (fetch) {
+    return function (...args) {
+      return fetchInterceptor(fetch, ...args);
+    };
+  })(env.fetch);
+}
+
+function runRefreshTokenPromise() {
+  refreshAccessTokenPromise = new Promise((resolve, reject) => {
+    // fetch new token with refresh token
+    return Promise.resolve(config.prepareRefreshTokenRequest(tokens.refreshToken))
+      .then(tokenRequest => fetch(tokenRequest))
+      .then(response => {
+        refreshAccessTokenPromise = null;
+
+        if (response.status !== STATUS_OK) {
+          throw new Error(ERROR_REFRESH_TOKEN_EXPIRED);
```

Comment:
Checked, I should only try to handle 401 and behave as normal fetch in all other cases, so if it throws, we rethrow. No magic.